# Python Functions: Medical Insurance Project
You are curious about how certain factors such as age, sex, BMI, number of children, and smoking status contribute to medical insurance costs.

You will apply your new knowledge of Python functions to write a useful function that calculates medical insurance costs.

Let’s get started!

## Creating a function

First, take a look at the code in script.py.

In this code, we estimate the medical insurance costs for two individuals, Maria and Omar, based on five variables:

- `age`: age of the individual in years
- `sex`: 0 for female, 1 for male
- `bmi`: individual’s body mass index
- `num_of_children`: number of children the individual has
- `smoker`: 0 for a non-smoker, 1 for a smoker

These variables are used in the following formula to estimate an individual’s insurance cost (in USD):

`insurance_cost = 250∗age − 128∗sex + 370∗bmi + 425∗num_of_children + 24000∗smoker − 12500`

Click “Save” to see the estimated insurance costs for Maria and Omar. 

In [12]:
# Initial variables for Maria 
age = 28
sex = 0  
bmi = 26.2
num_of_children = 3
smoker = 0  

# Estimate Maria's insurance cost
insurance_cost = 250*age - 128*sex + 370*bmi + 425*num_of_children + 24000*smoker - 12500

print("The estimated insurance cost for Maria is " + str(insurance_cost) + " dollars.")

# Initial variables for Omar
age = 35
sex = 1 
bmi = 22.2
num_of_children = 0
smoker = 1  

# Estimate Omar's insurance cost 
insurance_cost = 250*age - 128*sex + 370*bmi + 425*num_of_children + 24000*smoker - 12500

print("The estimated insurance cost for Omar is " + str(insurance_cost) + " dollars.")

The estimated insurance cost for Maria is 5469.0 dollars.
The estimated insurance cost for Omar is 28336.0 dollars.


The code used to estimate insurance costs for Maria and Omar looks quite similar – in both cases we calculate the insurance cost using the same formula and then print the output.

This code is a great candidate for a function because it involves repeating almost identical commands in multiple places.

Let’s start by defining a function called `calculate_insurance_cost()`. For now, your function should not have any parameters or output.

In [13]:
def calculate_insurance_cost():

SyntaxError: unexpected EOF while parsing (<ipython-input-13-fb6b7ccdfb75>, line 1)

Let’s outline the behavior we want our function to have. Inside of `calculate_insurance_cost()`, do the following:
- Create a variable called `estimated_cost`. For now, set this variable equal to a value of `1000`. You’ll add the full formula in the next step.
- Add a print statement that prints `estimated_cost`. You should output a message similar to: `"The estimated insurance cost for this person is xxx dollars."`
- Return `estimated_cost`

In [14]:
def calculate_insurance_cost():
    estimated_cost = 1000
    print('The estimated insurance cost for this person is ' + str(estimated_cost) + ' dollars.')
    return estimated_cost

## Adding parameters to a function
Nice job – you’ve created a simple Python function that we’ll use to estimate medical insurance costs.

However, the function currently returns a value of 1000. We want it to return our insurance cost formula instead.

Modify the function definition so that it contains five parameters:
- age
- sex
- bmi
- num_of_children
- smoker

Take a look at the hint if you need a reminder on how to add parameters to a function definition. 

In [15]:
def calculate_insurance_cost(age, sex, bmi, num_of_children, smoker):
    estimated_cost = 1000
    print('The estimated insurance cost for this person is ' + str(estimated_cost) + ' dollars.')
    return estimated_cost

Now that we have set up the function to take inputs for each of the values needed in the insurance formula, we can make use of them inside of our function.

In `calculate_insurance_cost()`, change the value of estimated_cost from 1000 to our formula for insurance cost.

Remember that the formula for insurance cost is: `250*age - 128*sex + 370*bmi + 425*num_of_children + 24000*smoker - 12500`

In [16]:
def calculate_insurance_cost(age, sex, bmi, num_of_children, smoker):
    estimated_cost = 250*age - 128*sex + 370*bmi + 425*num_of_children + 24000*smoker - 12500
    print('The estimated insurance cost for this person is ' + str(estimated_cost) + ' dollars.')
    return estimated_cost

## Calling a function
The function is now properly set up to calculate an individual’s medical insurance costs based on the five variables passed into it. Let’s test this out!

Go to the section of code that estimates Maria’s insurance cost.

- Rename `insurance_cost` as `maria_insurance_cost` and set it equal to `calculate_insurance_cost()` with the appropriate values for Maria as arguments. 
- Rename `insurance_cost` as `omar_insurance_cost` and set it equal to `calculate_insurance_cost()` with the appropriate values for Maria as arguments. 

In [17]:
# Estimate Maria's insurance cost
maria_insurance_cost = calculate_insurance_cost(28, 0, 26.2, 3, 0)

# Estimate Omar's insurance cost 
omar_insurance_cost = calculate_insurance_cost(35, 1, 22.2, 0, 1)

The estimated insurance cost for this person is 5469.0 dollars.
The estimated insurance cost for this person is 28336.0 dollars.


# Adding new parameters
Click “Save” to see the estimated insurance costs for Maria and Omar as calculated by your function.

In the output terminal, notice that it says `"The estimated insurance cost for this person is..."` but it does not specify the actual name of the person.

To fix this, begin by adding an additional parameter called `name` to the function definition. 

Next, modify the print statement in the function so that it includes the new `name` parameter, replacing `"this person"` with the actual `name` of the person.

In [18]:
def calculate_insurance_cost(age, sex, bmi, num_of_children, smoker, name):
    estimated_cost = 250*age - 128*sex + 370*bmi + 425*num_of_children + 24000*smoker - 12500
    print('The estimated insurance cost for ' + name + ' is ' + str(estimated_cost) + ' dollars.')
    return estimated_cost

# Estimate Maria's insurance cost
maria_insurance_cost = calculate_insurance_cost(28, 0, 26.2, 3, 0, 'Maria')

# Estimate Omar's insurance cost 
omar_insurance_cost = calculate_insurance_cost(35, 1, 22.2, 0, 1, name="Omar")

The estimated insurance cost for Maria is 5469.0 dollars.
The estimated insurance cost for Omar is 28336.0 dollars.


## Estimating your own insurance cost
In this example, we calculated the insurance costs for two individuals, but with our new code we can easily extend this to thousands or even millions of individuals.

To illustrate, estimate your own insurance cost.

At the bottom of your code, create a new `insurance_cost` variable for yourself, similar to how we did it for Maria and Omar.

Set the variable equal to a function call to `calculate_insurance_cost()`, passing in your own name, age, sex, bmi, number of children, and smoker status.

Click “Save” to see the result. 

In [19]:
# Estimate my's insurance cost 
my_insurance_cost = calculate_insurance_cost(35, 0, 22.2, 0, 0, name="Jane Doe")

The estimated insurance cost for Jane Doe is 4464.0 dollars.


## Extra
Congratulations! In this project, you created a useful function that calculates medical insurance costs based on the values passed into the function. Great job!

As a data scientist, you should always strive to make your code more clean and modular. Using functions to avoid repetitive code is a great way to do just that.

Now it’s your turn! If you’d like extra practice with Python functions, here are some suggestions to get you started:
- Modify the `calculate_insurance_cost()` function so that it returns two values – the output message and the estimated cost.
- Create a second function to calculate the difference between the insurance costs (given as inputs) of any two individuals and print a statement saying: `"The difference in insurance cost is xxx dollars."`

Happy coding! 

In [20]:
def calculate_insurance_cost(age, sex, bmi, num_of_children, smoker, name):
    estimated_cost = 250*age - 128*sex + 370*bmi + 425*num_of_children + 24000*smoker - 12500
    print(estimated_cost)
    print('The estimated insurance cost for ' + name + ' is ' + str(estimated_cost) + ' dollars.')
    return estimated_cost

# Estimate Maria's insurance cost
maria_insurance_cost = calculate_insurance_cost(28, 0, 26.2, 3, 0, 'Maria')

# Estimate Omar's insurance cost 
omar_insurance_cost = calculate_insurance_cost(35, 1, 22.2, 0, 1, name="Omar")

5469.0
The estimated insurance cost for Maria is 5469.0 dollars.
28336.0
The estimated insurance cost for Omar is 28336.0 dollars.


In [21]:
def diff_insurance_cost(cost1, cost2):
    diff_cost = cost1 - cost2
    print('The difference in insurance cost is ' + str(diff_cost) + ' dollars.')
    return diff_cost

diff_insurance_cost(maria_insurance_cost, omar_insurance_cost)

The difference in insurance cost is -22867.0 dollars.


-22867.0

Maria's insurance cost is 22867 dollars less than Omar's insurance cost